src/primitives.cr

```diff
@@ -49,6 +49,45 @@ class Reference
   @[Primitive(:object_id)]
   def object_id : UInt64
   end
+
+  # Performs basic initialization so that the given *address* is ready for use
+  # as an object's instance data. Returns *address* cast to `self`'s type.
+  #
+  # More specifically, this is the part of object initialization that occurs
+  # after memory allocation and before calling one of the `#initialize`
+  # overloads. It zeroes the memory, sets up the type ID (necessary for dynamic
+  # dispatch), and then runs all inline instance variable initializers.
+  #
+  # *address* must point to a suitably aligned buffer of at least
+  # `instance_sizeof(self)` bytes.
+  #
+  # WARNING: This method is unsafe, as it assumes the caller is responsible for
+  # managing the memory at the given *address* manually.
+  #
+  # ```
+  # class Foo
+  #   getter i : Int64
+  #   getter str = "abc"
+  #
+  #   def initialize(@i)
+  #   end
+  #
+  #   foo_buffer = LibC.malloc(24)
+  #   foo = pre_initialize(foo)
+  #   foo.i                  # => 0
+  #   foo.str                # => "abc"
+  #   (foo || "").is_a?(Foo) # => true
+  #
+  #   foo.initialize(123_i64)
+  #   foo.i # => 123
+  # end
```

Comment:
Looks like the nesting is off:

```suggestion
  # end
  #
  # foo_buffer = LibC.malloc(24)
  # foo = Foo.pre_initialize(foo)
  # foo.i                  # => 0
  # foo.str                # => "abc"
  # (foo || "").is_a?(Foo) # => true
  #
  # foo.initialize(123_i64)
  # foo.i # => 123
```

Reply:
It's quite uncommon though, to put this kind of top-level code in a type scope. IMO it would be easier to understand the code if was moved outside of `Foo`. I suppose we can leave it there due to the complication with `#initialize`'s scope. At least I'd add the type as receiver of `pre_initialize` though, for clarity.